### PR TITLE
[Locale] Change language negotiation to RFC 4647 based

### PR DESCRIPTION
**Pull request by gseric**
| Q               | A
| --------------- | -----
| Branch?         | 1.11
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| License         | MIT

Following this PR: #13708

### CODE REVIEW

**Comment by lchrusciel** [comment]
Hey Goran,

thanks a lot for your contribution and welcome among our contributors :). I will try to review it ASAP (within two weeks probably).

File: tests/Api/Shop/ProductsTest.php [hunk 1]
```diff
@@ -63,7 +63,7 @@ public function it_returns_product_with_translations_in_locale_from_header(): vo
             sprintf('/api/v2/shop/products/%s', $product->getCode()),
             [],
             [],
-            ['CONTENT_TYPE' => 'application/ld+json', 'HTTP_ACCEPT' => 'application/ld+json', 'HTTP_ACCEPT_LANGUAGE' => 'de_DE']
+            ['CONTENT_TYPE' => 'application/ld+json', 'HTTP_ACCEPT' => 'application/ld+json', 'HTTP_ACCEPT_LANGUAGE' => 'de-DE']
```
Comment (lchrusciel) [inline tests/Api/Shop/ProductsTest.php:66]: Can we duplicate this test with two different accept-language headers?

File: src/Sylius/Bundle/LocaleBundle/Resources/config/services.xml [hunk 2]
```diff
@@ -42,7 +42,7 @@
         <service id="Sylius\Bundle\LocaleBundle\Context\RequestHeaderBasedLocaleContext">
             <argument type="service" id="request_stack" />
             <argument type="service" id="sylius.locale_provider" />
-            <tag name="sylius.context.locale" priority="256" />
+            <tag name="sylius.context.locale" priority="32" />
```
Comment (lchrusciel) [inline src/Sylius/Bundle/LocaleBundle/Resources/config/services.xml:45]: This change seems to be irrelevant to the main purpose of this PR. Is it related to: https://github.com/Sylius/Sylius/pull/13708#issuecomment-1095396785?

File: src/Sylius/Bundle/LocaleBundle/Context/RequestHeaderBasedLocaleContext.php [hunk 3]
```diff
@@ -33,17 +38,21 @@ public function getLocaleCode(): string
...
-            throw new LocaleNotFoundException('No locale is set on the master request\'s headers.');
-        }
-
         $availableLocalesCodes = $this->localeProvider->getAvailableLocalesCodes();
-        if (!in_array($localeCode, $availableLocalesCodes, true)) {
-            throw LocaleNotFoundException::notAvailable($localeCode, $availableLocalesCodes);
+
+        // Request::getPreferredLanguage() returns first available locale code if none matches. To allow detection of
+        // this unwanted behavior, we will prepend special locale code to the list of available locale codes.
+        $prependedAvailableLocalesCodes = array_merge(['FIRSTLOCALECODE'], $availableLocalesCodes);
```
Comment (lchrusciel) [inline src/Sylius/Bundle/LocaleBundle/Context/RequestHeaderBasedLocaleContext.php:45]: What would you say for something more meaningful?
```suggestion
        $prependedAvailableLocalesCodes = array_merge(['ALL_CODES_ARE_INVALID_STUB'], $availableLocalesCodes);
```

+ can you put this value into private consts?

**Review by lchrusciel — COMMENTED** [review]
Hey Goran,

thanks for the PR + pretty nice docs update. You rock. I've added a few review comments

**Review by Zales0123 — COMMENTED** [review]
👍 but we should probably also mention this change in the UPGRADE file

**Review by jakubtobiasz — APPROVED** [review]

**Comment by jakubtobiasz** [comment]
Thank you, @gseric!

**Comment by jakubtobiasz** [comment]
and @NoResponseMate too!
